### PR TITLE
[SYCL] Mark kernel object as invalid if incorrect accessor format

### PR DESCRIPTION
**Pull request by elizabethandrews**
If kernel object is marked invalid, we do not continue processing
it, i.e. the compiler does not proceed to generate the corresponding
openCL kernel. Invalid captures/kernel object fields should cause
kernel object to be marked as invalid. This patch fixes a bug where
invalid accessors did not do so.

Part 1 of fix for  - https://github.com/intel/llvm/issues/4115

Signed-off-by: Elizabeth Andrews <elizabeth.andrews@intel.com>

### CODE REVIEW

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -443,7 +443,7 @@ static void checkSYCLType(Sema &S, QualType Ty, SourceRange Loc,
   // for example the quad type __float128 will cause errors in the
   // SPIR-V translation phase.
   // Here we check any potentially unsupported declaration and issue
-  // a deferred diagnostic, which will be emitted iff the declaration
+  // a deferred diagnostic, which will be emitted if the declaration
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:446]: Was this change intended? I read the original "iff" as "if and only iff" (https://www.merriam-webster.com/dictionary/iff).

File: clang/lib/Sema/SemaSYCL.cpp [hunk 2]
```diff
@@ -1605,71 +1605,71 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
...
+        return SemaRef.Diag(
+                   Loc,
+                   diag::err_sycl_invalid_accessor_property_list_template_param)
+               << /*accessor_property_list pack argument*/ 1 << /*type*/ 1;
       QualType PropTy = Prop->getAsType();
-      if (Util::isSyclBufferLocationType(PropTy))
-        checkBufferLocationType(PropTy, Loc);
+      if (Util::isSyclBufferLocationType(PropTy) &&
+          checkBufferLocationType(PropTy, Loc))
+        return true;
```
[the remark this answers is not in the document]
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:1642]: This is inside a loop. You shouldn't exit loop early if false.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 1 above)
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp]: Oh right. My bad. I didn't think of 'if and only if'. Reverted this change. Thanks!

File: clang/lib/Sema/SemaSYCL.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (premanandrao) [reply clang/lib/Sema/SemaSYCL.cpp:1642]: Ah!  Got it.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 1 above)
Comment (keryell) [reply clang/lib/Sema/SemaSYCL.cpp]: This is a common recurring issue: what about not using abbreviations in comments, considering the price of storage for 1 character nowadays? :-)

Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp]: I'm ok with your suggestion and will keep that in mind for my patches :) However, I'm not sure about enforcing it during review since I'm not sure what clang community's rule is . I suspect we just go along with whatever community standard is. @AaronBallman might know the answer

Comment (AaronBallman) [reply clang/lib/Sema/SemaSYCL.cpp]: We won't be getting rid of abbreviations in general any time soon (AST, CFG, IR, LLVM, HIP, CUDA, etc), but `iff` is a bit different from those IMO. I looked through the Clang codebase and found about 100 uses of `iff` that all seem to be correct rather than typos (from spot checking, so I might have missed some). Somewhat amusingly `if and only if` appears... about 100 times as well. :-D

I don't have a strong opinion and I doubt community does either. I'm fine leaving this as `iff`, but @elizabethandrews is free to expand it to the long form it iff she thinks it's worthwhile. :-)

**Review by AaronBallman — APPROVED** [review]
LGTM!

File: clang/lib/Sema/SemaSYCL.cpp (hunk 1 above)
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp]: Since this comment is technically in an area I don't need to touch in this patch, I'm going to leave it as is for now. I'll keep an eye out in the future :)

Comment (keryell) [reply clang/lib/Sema/SemaSYCL.cpp]: Good comments means empathy, inclusive coding, a social experience...
So it makes sense to have less abbreviations. :-)
